### PR TITLE
Avoid unnecessary API update and re-enqueue action

### PR DESCRIPTION
**Pull request by wangyysde**
fix issue #3490 

ref: #3502

### CODE REVIEW

**Comment by lowang-bh** [comment]
please rename the PR's title with some meaningful text.

**Comment by wangyysde** [comment]
> please rename the PR's title with some meaningful text.

done

File: pkg/controllers/job/job_controller_actions.go [hunk 1]
```diff
@@ -465,6 +470,11 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
 		RetryCount:          job.Status.RetryCount,
 	}
 
+	if reflect.DeepEqual(newStatus, job.Status) {
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:473]: We should call `updateStatus` first before compare.

File: pkg/controllers/job/job_controller_actions.go [hunk 2]
```diff
@@ -137,6 +139,9 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
 
 	// Update Job status
 	newJob, err := cc.vcClient.BatchV1alpha1().Jobs(job.Namespace).UpdateStatus(context.TODO(), job, metav1.UpdateOptions{})
+	if errors.IsNotFound(err) {
+		return nil
```
Comment (lowang-bh) [inline pkg/controllers/job/job_controller_actions.go:143]: Let's record a log here.

File: pkg/controllers/job/job_controller_actions.go (hunk 1 above)
Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go]: ok. done

File: pkg/controllers/job/job_controller_actions.go (hunk 2 above)
Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go:145]: ok. done

File: pkg/controllers/job/job_controller_actions.go [hunk 3]
```diff
@@ -137,6 +139,11 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
 
 	// Update Job status
 	newJob, err := cc.vcClient.BatchV1alpha1().Jobs(job.Namespace).UpdateStatus(context.TODO(), job, metav1.UpdateOptions{})
+	if errors.IsNotFound(err) {
+		klog.Errorf("Job %v/%v was not found",
```
Comment (lowang-bh) [inline pkg/controllers/job/job_controller_actions.go:143]: Keep code just one line is ok.

Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go]: done

File: pkg/controllers/job/job_controller_actions.go [hunk 4]
```diff
@@ -464,12 +471,18 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
 		Conditions:          job.Status.Conditions,
 		RetryCount:          job.Status.RetryCount,
 	}
+	oldStatus := job.Status
+	job.Status = newStatus
 
 	if updateStatus != nil && updateStatus(&job.Status) {
 		job.Status.State.LastTransitionTime = metav1.Now()
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:478]: line 478-480 should be executed after `reflect.DeepEqual` check result is false, or line 483 check will always be true.

Comment (Monokaix) [reply pkg/controllers/job/job_controller_actions.go:478]:
``` go
if updateStatus != nil {
  updateStatus(&newStatus)
}

if reflect.DeepEqual(oldStatus, newStatus) {
  klog.xxx
  return nil
}else{
job.Status = newStatus
job.Status.State.LastTransitionTime = metav1.Now()
jobCondition = newCondition(job.Status.State.Phase, &job.Status.State.LastTransitionTime)
job.Status.Conditions = append(job.Status.Conditions, jobCondition)
}

```
and remove line 474-475

Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go]: done. Thks

**Review by william-wang — APPROVED** [review]
/approve

File: pkg/controllers/job/job_controller_actions.go [hunk 5]
```diff
@@ -465,7 +472,15 @@ func (cc *jobcontroller) syncJob(jobInfo *apis.JobInfo, updateStatus state.Updat
...
 
-	if updateStatus != nil && updateStatus(&job.Status) {
+	if updateStatus != nil {
+		updateStatus(&newStatus)
+	}
+
+	if reflect.DeepEqual(job.Status, newStatus) {
+		klog.V(3).Infof("Job <%s/%s> has not updated for no changing", job.Namespace, job.Name)
+		return nil
+	} else {
```
Comment (lowang-bh) [inline pkg/controllers/job/job_controller_actions.go:482]: Let's remove the else.

Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go]: done

Comment (lowang-bh) [reply pkg/controllers/job/job_controller_actions.go]: I mean code in else can be out of if else case. The `else {` is not nessasary any more.

**Comment by lowang-bh** [comment]
> /retest

please check the last comment @wangyysde

File: pkg/controllers/job/job_controller_actions.go (hunk 5 above)
Comment (wangyysde) [reply pkg/controllers/job/job_controller_actions.go]: ok, I have done

**Comment by wangyysde** [comment]
@lowang-bh Hi, could you set retest to this PR? I found the following message in E2E test log: 
 is forbidden: unable to create new content in namespace c800v8rb because it is being terminated", Reason:"Forbidden", Details:(*v1.StatusDetails)(0xc000bc1d40), Code:403}

**Comment by lowang-bh** [comment]
> @lowang-bh Hi, could you set retest to this PR? 

You can close it and reopen, it will trigger ci.

**Comment by googs1025** [comment]
> @lowang-bh Hi, could you set retest to this PR? I found the following message in E2E test log: is forbidden: unable to create new content in namespace c800v8rb because it is being terminated", Reason:"Forbidden", Details:(*v1.StatusDetails)(0xc000bc1d40), Code:403}

another way: 
`git commit --amend -s --no-edit` and push again

**Comment by wangyysde** [comment]
@googs1025 ok. thks

**Review by william-wang — APPROVED** [review]
/approve

**Pull request merged**